Use quote function for all errors/warnings

Marenz opened the pull request.

Marenz committed:
Use quote function for all errors/warnings

leonardoalt commented on libsolidity/formal/BMC.cpp:
```diff
@@ -764,7 +764,8 @@ void BMC::checkCondition(
 	{
 		std::ostringstream message;
-		message << _description << " happens here";
+		message << _description << " happens here.";
```
I'm not sure about this one. The idea is that if there is a counterexample the sentence goes on as `...happens here for:`, where in this case it'll be `happens here. for:`

Marenz replied:
I assumed that's your intention but you are actually never modifying the `message` variable again in your first `if`-branch. You're using it as-is. Instead you add your example as a `SecondarySourceLocation`. Not sure if you intended it that way.

leonardoalt replied:
Right, because text is just added after `message`. But in the `else` branch, where there are no counterexamples, `.` is added to `message`, that's why it doesn't have `.` before branching.

Marenz replied:
No.. it's not. Text is _not_ added after `message`. Only `SecondarySourceLocation` "text". If the first `if` branch is taken, the `.` will be missing completely.

Marenz replied:
Most of your tests show this, too ;) As an example, see 
https://github.com/ethereum/solidity/blob/develop/test/libsolidity/smtCheckerTests/functions/constructor_simple.sol#L16
found using `git grep "happens here$"` which results in about 360 tests.
And only two tests with the `.`

chriseth commented on liblangutil/ErrorReporter.cpp:
```diff
@@ -28,6 +28,35 @@ using namespace std;
+{
+	if (
+		_message.size() == 0 ||
+		_message[_message.size()-1] == '.' ||
```
style

chriseth replied:
actually multiple: binary operators have spaces around them and you should use `_message.back()`.

Marenz replied:
`_message.back()` is certainly nicer. I don't understand the spaces part though? the `||` operators do have a space before the `or`ed conditions and we don't want a trailing spaces as far as I know

chriseth replied:
It was about the `-`

chriseth commented on liblangutil/ErrorReporter.cpp:
```diff
@@ -28,6 +28,35 @@ using namespace std;
+
+}
+
+void checkMessage(string const _message)
```
Why do you pass it by value and then don't move below?

Marenz replied:
should be by reference. Had a too long break of programming ;)

chriseth commented on libsolidity/analysis/NameAndTypeResolver.cpp:
```diff
@@ -98,13 +98,13 @@ bool NameAndTypeResolver::performImports(SourceUnit& _sourceUnit, map<string, So
-							imp->path() +
-							"\")."
+							"Declaration" +
+							quote(alias.symbol->name()) +
```
I wouldn't have guessed that adding spaces around would be the default implementation.

Marenz replied:
I tried to use the most common case as default. Mabye using a different functionname instead is better? `quoteSpace` for example?

chriseth replied:
What is the advantage of implicit spaces?

Marenz replied:
I felt `"This is an" + quote("example") + "message."` reads nicer than
`"This is an " + quote("example") + " message."`. Basically you need to almost always add those spaces extra and I tried to save the extra grunt work.

Marenz replied:
Especially with multiple quotes in one sentence:
`"Elements" + quote("a") + "and" + quote("b") + "cannot be used in this context."`

Marenz replied:
it's not a terrible issue and if you prefer I can just go with a normal `quote()` that does only quote and nothing else

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -188,23 +192,23 @@ bool SyntaxChecker::visit(Continue const& _continueStatement)
 	if (m_inLoopDepth <= 0)
 		// we're not in a for/while loop, report syntax error
-		m_errorReporter.syntaxError(_breakStatement.location(), "\"break\" has to be in a \"for\" or \"while\" loop.");
+		m_errorReporter.syntaxError(_breakStatement.location(), quote("break", false) + " has to be in a" + quote("for") + "or" + quote("while") + "loop.");
```
Is this really more readable?

Marenz replied:
because of the `, false`? or generally? Note that it's not only about readability, but also that we can easily change the quotes used depending on the output, e.g. fancy terminals might support quotes like `»this«`

Marenz commented:
I removed the second commit for now (couldn't get some tests to work in a reasonable time) and rebased on latest develop.

the second commit can be found in the branch `terminate-with-dot`

Marenz commented:
Ah and of course, I added as discussed a second function `quoteSpace()` to make the usage more intuitive.

chriseth commented on Changelog.md:
```diff
@@ -8,6 +8,7 @@ Compiler Features:
  * Yul Optimizer: Apply penalty when trying to rematerialize into loops.
 
 Bugfixes:
+ * Made usage of quotes in error and warning messages consistent
```
```suggestion
 * Consistent use of quotes in error and warning messages.
```

chriseth commented on liblangutil/ErrorReporter.cpp:
```diff
@@ -28,6 +28,20 @@ using namespace std;
 using namespace solidity;
 using namespace solidity::langutil;
 
+namespace solidity::langutil {
```
Usually we use explicit namespace prefixes if the reference is ambiguous instead of opening namespaces in cpp files.

chriseth commented on liblangutil/ErrorReporter.cpp:
```diff
@@ -28,6 +28,20 @@ using namespace std;
+	return "\"" + _name + "\"";
+}
+
+std::string quoteSpace(std::string const& _name)
```
```suggestion
std::string quoteSpaced(std::string const& _name)
```
?

chriseth commented on liblangutil/ErrorReporter.cpp:
```diff
@@ -28,6 +28,20 @@ using namespace std;
+
+std::string quoteSpace(std::string const& _name)
+{
+	return " " +quote(_name) + " ";
```
```suggestion
	return " " + quote(_name) + " ";
```
Please also make an addition to the style checker.

chriseth commented on liblangutil/ParserBase.cpp:
```diff
@@ -65,14 +65,14 @@ string ParserBase::tokenName(Token _token)
 		return "end of source";
 	else if (TokenTraits::isReservedKeyword(_token))
-		return "reserved keyword '" + TokenTraits::friendlyName(_token) + "'";
+		return "reserved keyword " + quote(TokenTraits::friendlyName(_token));
```
Does that cause any breaking changes? This is not only used in error messages.

chriseth commented on libsolidity/analysis/ContractLevelChecker.cpp:
```diff
@@ -228,8 +228,8 @@ void ContractLevelChecker::checkAbstractFunctions(ContractDefinition const& _con
 		m_errorReporter.typeError(_contract.location(), ssl,
-			"Contract \"" + _contract.annotation().canonicalName
-			+ "\" should be marked as abstract.");
+			"Contract" + quoteSpace(_contract.annotation().canonicalName)
```
style

chriseth commented on libsolidity/analysis/ContractLevelChecker.cpp:
```diff
@@ -368,7 +368,7 @@ void ContractLevelChecker::checkHashCollisions(ContractDefinition const& _contra
 			m_errorReporter.typeError(
 				_contract.location(),
-				string("Function signature hash collision for ") + it.second->externalSignature()
+				string("Function signature hash collision for ") + quote(it.second->externalSignature()) + "."
```
No `string(...)` needed anymore.

chriseth commented on libsolidity/analysis/ContractLevelChecker.cpp:
```diff
@@ -425,11 +425,11 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
-			std::string("Contract \"") +
-			_contract.name() +
-			"\" does not use ABIEncoderV2 but wants to inherit from a contract " +
+			std::string("Contract") +
```
Same here.

chriseth commented on libsolidity/analysis/OverrideChecker.cpp:
```diff
@@ -699,9 +699,9 @@ void OverrideChecker::checkAmbiguousOverridesInternal(set<OverrideProxy> _baseCa
-		"\". Two or more base classes define " + callableName + " with same " + distinguishigProperty + ".";
+		"Derived contract must override " + callableName + " " +
+		quote(_baseCallables.begin()->name()) +
+		". Two or more base classes define " + callableName + " with same " + distinguishigProperty + ".";
```
Why is `callableName` not quoted?

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -60,13 +60,16 @@ void SyntaxChecker::endVisit(SourceUnit const& _sourceUnit)
-				to_string(recommendedVersion.patch()) +
-				string(";\"");
+				" Consider adding " +
+				quoteSpace("pragma solidity ^" +
```
```suggestion
				quoteSpace(
				    "pragma solidity ^" +
```

chriseth commented on test/libsolidity/syntaxTests/nameAndTypeResolution/551_warn_about_address_members_on_non_this_contract_callcode.sol:
```diff
@@ -5,4 +5,4 @@ contract C {
 }
 // ----
-// TypeError: (65-75): Member "callcode" not found or not visible after argument-dependent lookup in contract C. Use "address(c).callcode" to access this address member.
+// TypeError: (65-75): Member "callcode" not found or not visible after argument-dependent lookup in "contract C". Use "address(c).callcode" to access this address member.
```
Do we really want to quote `contract C`?

chriseth commented on test/libsolidity/syntaxTests/nameAndTypeResolution/351_callcode_deprecated.sol:
```diff
@@ -4,4 +4,4 @@ contract test {
 }
 // ----
-// TypeError: (55-77): "callcode" has been deprecated in favour of "delegatecall".
+// TypeError: (55-77): "callcode" has been deprecated in favour of"delegatecall".
```
space

chriseth commented:
Even though it is "finished", I fear this still draws several hours of work from the team. Not sure if we should continue.

chriseth commented:
I reviewed less than 10% and already added 10 comments. This will not end well :)

chriseth commented:
Added "to discuss" so we talk about what to do with this on monday.

erak commented:
The overall conclusion was, to rather have a Whiskers-like approach to more readable and consistent error reporting. Closing this in favour of https://github.com/ethereum/solidity/issues/8218.

erak commented:
The overall conclusion was, to rather have a Whiskers-like approach to more readable and consistent error reporting. Closing this in favour of #8218.
